src/LanguageServer/Impl/CodeActions/MissingImportCodeActionProvider.cs

```diff
@@ -0,0 +1,635 @@
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Python.Analysis;
+using Microsoft.Python.Analysis.Analyzer;
+using Microsoft.Python.Analysis.Analyzer.Expressions;
+using Microsoft.Python.Analysis.Core.DependencyResolution;
+using Microsoft.Python.Analysis.Core.Interpreter;
+using Microsoft.Python.Analysis.Diagnostics;
+using Microsoft.Python.Analysis.Modules;
+using Microsoft.Python.Analysis.Types;
+using Microsoft.Python.Analysis.Values;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.Collections;
+using Microsoft.Python.Core.Text;
+using Microsoft.Python.LanguageServer.Diagnostics;
+using Microsoft.Python.LanguageServer.Indexing;
+using Microsoft.Python.LanguageServer.Protocol;
+using Microsoft.Python.Parsing.Ast;
+using Range = Microsoft.Python.Core.Text.Range;
+
+namespace Microsoft.Python.LanguageServer.CodeActions {
+    internal sealed class MissingImportCodeActionProvider : ICodeActionProvider {
+        private const int ModuleLoadTimeout = 10 * 1000; // 10 seconds
+
+        public static readonly ICodeActionProvider Instance = new MissingImportCodeActionProvider();
+
+        // right now, it is a static. in future, we might consider giving an option to users to customize this list
+        // also, right now, it is text based. so if module has same name, they will get same suggestion even if
+        // the module is not something user expected
+        private static readonly Dictionary<string, string> WellKnownAbbreviationMap = new Dictionary<string, string>() {
+            { "numpy", "np" },
+            { "pandas", "pd" }
+        };
+
+        private MissingImportCodeActionProvider() {
+        }
+
+        public ImmutableArray<string> FixableDiagnostics => ImmutableArray<string>.Create(
+            ErrorCodes.UndefinedVariable, ErrorCodes.VariableNotDefinedGlobally, ErrorCodes.VariableNotDefinedNonLocal);
+
+        public async Task<IEnumerable<CodeAction>> GetCodeActionAsync(IDocumentAnalysis analysis, DiagnosticsEntry diagnostic, CancellationToken cancellationToken) {
+            var finder = new ExpressionFinder(analysis.Ast, FindExpressionOptions.Complete);
+            var node = finder.GetExpression(diagnostic.SourceSpan);
+            if (!(node is NameExpression)) {
+                return Enumerable.Empty<CodeAction>();
+            }
+
+            var interpreter = analysis.Document.Interpreter;
+            var pathResolver = interpreter.ModuleResolution.CurrentPathResolver;
+
+            var name = node.ToCodeString(analysis.Ast);
```

Comment:
`node is NameExpression nex` (above), then nex.Name. `ToCodeString` is buggy and only used in tests.

Reply:
thank you.